feature/typed_var

nitely opened the pull request:
This feature was described in comments of #2 

* Adds var validators
* Adds support for I10L URLs, mostly coz validating on a encoded string does not make much sense
* Adds support for clashing URLs
* *Breaking change*: on clashing URLs first added URL will match. Previous to this last added URL would had match.

Usage:

```python
is_num = lambda part: part.isdigit()
routes.add('api/user/:id', {'GET': my_get_controller}, {'id': is_num})
route = routes.match('api/user/123')
route.params
# {'id': '123'}

route = routes.match('api/user/bad')
# Raises RouteError
```

nitely committed:
pattern var validate

nitely committed:
fix docstring

nitely committed:
fix DRY typed vars tests

nitely committed:
backtracking tests

nitely commented:
Leaving this open since it has some missing functionality (the pattern matching). To fix that it's basically a matter of changing the `_Route` tree end into a list of routes and going through them until one matches.

yohanboniface commented:
Nice move!

I'd personally decouplate route and validation, as per API, so something like:

```python
routes = Routes()
routes.validate(int=my_in_validator, foobar=my_foobar_validator)
routes.add('/test/int:id/foobar:param')
```

nitely committed:
support I10L URLs

nitely committed:
support clashing URLs

nitely committed:
fix use unquote instead of unquote_plus

plus signs should not be treated as spaces within the components of a URL (only within a querystring)

nitely committed:
I10L tests

nitely commented:
@yohanboniface yeah, I was planning on supporting both cases but I think I'll support just the one you suggested.

nitely committed:
add default validator for naked vars

nitely committed:
refactor

nitely committed:
improve docs

nitely committed:
max_depth -> 10

ahopkins commented:
Just curious ... what is the status of this? Is this project dead?

nitely commented:
It's completed (since inception). I may or may not add ~~bloat~~ new features to it. This PR was not merged 'cause I never made my mind about #7
